File: Data Structures & Algorithms/top-k-elements-in-list/submission-3.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {number[]}
     */
    topKFrequent(nums, k) {

const freq = new Map();
for (const n of nums) freq.set(n, (freq.get(n) || 0) + 1);


const buckets = Array.from({ length: nums.length + 1 }, () => []);



for (const [num, count] of freq) {
    console.log(num, count)
    buckets[count].push(num);
}

console.log("buckets", buckets);
let res = [];
for (let i = buckets.length - 1; i >= 1 && res.length < k; i--) {
  res.push(...buckets[i]);
}

return res;

    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-3.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int len = nums.length;
        int[] res = new int[len];
        int[] l = new int[len];
        int[] r = new int[len];
        if(len<=1){
            return nums;
        }
         l[0] = nums[0];
         r[len-1] = nums[len-1];
        for(int i=1; i<len; i++){
            l[i] = l[i-1] * nums[i];
        }
        for(int i=len-2; i>=0; i--){
            r[i] = r[i+1] * nums[i];
        }

        res[0]= r[1];
        res[len-1]= l[len-2];
        for(int i=1; i<len-1; i++){
            res[i] = l[i-1]*r[i+1];
        }
        return res;
    }
}  
